### PR TITLE
[RFC&PI] Thread mul multi mod

### PR DESCRIPTION
**Pull request by tthsqe12**
The threading helps the quadratic part too.

### CODE REVIEW

File: fmpz/get_nmod.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <https://www.gnu.org/licenses/>.
+*/
+
+#include "fmpz.h"
+
+mp_limb_t fmpz_get_nmod(const fmpz_t aa, nmod_t mod)
```
Comment (wbhart) [inline fmpz/get_nmod.c:14]: How is this different to fmpz_fdiv_ui?

**Comment by wbhart** [comment]
Here are the new timings:

```
dim  16: min 19.471 ns  max 24.963 ns  aver 22.550 ns
dim  23: min 11.457 ns  max 18.430 ns  aver 15.027 ns
dim  30: min 6.420 ns  max 12.840 ns  aver 10.288 ns
dim  38: min 4.433 ns  max 8.866 ns  aver 7.142 ns
dim  47: min 3.853 ns  max 8.187 ns  aver 6.341 ns
dim  57: min 2.945 ns  max 7.854 ns  aver 5.727 ns
dim  69: min 2.609 ns  max 6.088 ns  aver 4.276 ns
dim  82: min 2.267 ns  max 4.534 ns  aver 3.476 ns
dim  97: min 1.826 ns  max 3.652 ns  aver 2.861 ns
dim 114: min 1.350 ns  max 3.375 ns  aver 2.587 ns
dim 133: min 1.700 ns  max 2.975 ns  aver 2.338 ns
dim 154: min 1.643 ns  max 2.738 ns  aver 2.145 ns
dim 178: min 1.241 ns  max 2.128 ns  aver 1.714 ns
dim 205: min 1.277 ns  max 1.857 ns  aver 1.509 ns
dim 235: min 0.771 ns  max 1.464 ns  aver 1.220 ns
dim 269: min 0.668 ns  max 1.438 ns  aver 1.062 ns
dim 307: min 0.588 ns  max 1.210 ns  aver 0.962 ns
dim 350: min 0.676 ns  max 1.259 ns  aver 0.980 ns
dim 398: min 0.650 ns  max 1.126 ns  aver 0.933 ns
dim 452: min 0.466 ns  max 0.975 ns  aver 0.780 ns
dim 513: min 0.474 ns  max 0.948 ns  aver 0.758 ns
dim 582: min 0.421 ns  max 0.878 ns  aver 0.697 ns
dim 659: min 0.405 ns  max 0.790 ns  aver 0.647 ns
dim 746: min 0.889 ns  max 1.260 ns  aver 1.069 ns
dim 844: min 0.956 ns  max 1.944 ns  aver 1.547 ns
dim 954: min 0.909 ns  max 1.790 ns  aver 1.441 ns
dim 1078: min 0.849 ns  max 1.656 ns  aver 1.333 ns
dim 1217: min 0.787 ns  max 1.515 ns  aver 1.220 ns
dim 1374: min 0.737 ns  max 1.391 ns  aver 1.142 ns
dim 1550: min 0.712 ns  max 1.344 ns  aver 1.084 ns
dim 1748: min 0.674 ns  max 1.228 ns  aver 1.009 ns
dim 1971: min 0.624 ns  max 1.211 ns  aver 0.979 ns
```

Here are the times for the old fmpz_mat_mul:

```
dim  16: min 19.471 ns  max 24.963 ns  aver 22.550 ns
dim  23: min 11.457 ns  max 18.430 ns  aver 15.027 ns
dim  30: min 6.420 ns  max 12.840 ns  aver 10.288 ns
dim  38: min 4.433 ns  max 8.866 ns  aver 7.142 ns
dim  47: min 3.853 ns  max 8.187 ns  aver 6.341 ns
dim  57: min 2.945 ns  max 7.854 ns  aver 5.727 ns
dim  69: min 2.609 ns  max 6.088 ns  aver 4.276 ns
dim  82: min 2.267 ns  max 4.534 ns  aver 3.476 ns
dim  97: min 1.826 ns  max 3.652 ns  aver 2.861 ns
dim 114: min 1.350 ns  max 3.375 ns  aver 2.587 ns
dim 133: min 1.700 ns  max 2.975 ns  aver 2.338 ns
dim 154: min 1.643 ns  max 2.738 ns  aver 2.145 ns
dim 178: min 1.241 ns  max 2.128 ns  aver 1.714 ns
dim 205: min 1.277 ns  max 1.857 ns  aver 1.509 ns
dim 235: min 0.771 ns  max 1.464 ns  aver 1.220 ns
dim 269: min 0.668 ns  max 1.438 ns  aver 1.062 ns
dim 307: min 0.588 ns  max 1.210 ns  aver 0.962 ns
dim 350: min 0.676 ns  max 1.259 ns  aver 0.980 ns
dim 398: min 0.650 ns  max 1.126 ns  aver 0.933 ns
dim 452: min 0.466 ns  max 0.975 ns  aver 0.780 ns
dim 513: min 0.474 ns  max 0.948 ns  aver 0.758 ns
dim 582: min 0.421 ns  max 0.878 ns  aver 0.697 ns
dim 659: min 0.405 ns  max 0.790 ns  aver 0.647 ns
dim 746: min 0.889 ns  max 1.260 ns  aver 1.069 ns
dim 844: min 0.956 ns  max 1.944 ns  aver 1.547 ns
dim 954: min 0.909 ns  max 1.790 ns  aver 1.441 ns
dim 1078: min 0.849 ns  max 1.656 ns  aver 1.333 ns
dim 1217: min 0.787 ns  max 1.515 ns  aver 1.220 ns
dim 1374: min 0.737 ns  max 1.391 ns  aver 1.142 ns
dim 1550: min 0.712 ns  max 1.344 ns  aver 1.084 ns
dim 1748: min 0.674 ns  max 1.228 ns  aver 1.009 ns
dim 1971: min 0.624 ns  max 1.211 ns  aver 0.979 ns
```

And here are the times from my fmpz_mat_blas_tuning branch:

```
dim  16: min 7.489 ns  max 27.959 ns  aver 19.638 ns
dim  23: min 9.962 ns  max 28.891 ns  aver 20.174 ns
dim  30: min 10.370 ns  max 29.136 ns  aver 20.823 ns
dim  38: min 9.851 ns  max 19.702 ns  aver 14.448 ns
dim  47: min 6.261 ns  max 19.745 ns  aver 12.441 ns
dim  57: min 6.872 ns  max 17.181 ns  aver 12.272 ns
dim  69: min 3.914 ns  max 5.653 ns  aver 4.639 ns
dim  82: min 2.267 ns  max 5.894 ns  aver 3.930 ns
dim  97: min 2.922 ns  max 5.113 ns  aver 4.018 ns
dim 114: min 2.362 ns  max 3.375 ns  aver 3.037 ns
dim 133: min 1.700 ns  max 3.400 ns  aver 2.763 ns
dim 154: min 2.190 ns  max 3.286 ns  aver 2.647 ns
dim 178: min 1.596 ns  max 2.660 ns  aver 2.216 ns
dim 205: min 1.509 ns  max 2.438 ns  aver 2.051 ns
dim 235: min 1.310 ns  max 2.003 ns  aver 1.695 ns
dim 269: min 1.027 ns  max 1.901 ns  aver 1.438 ns
dim 307: min 1.037 ns  max 1.763 ns  aver 1.354 ns
dim 350: min 0.956 ns  max 1.563 ns  aver 1.294 ns
dim 398: min 0.999 ns  max 1.412 ns  aver 1.161 ns
dim 452: min 0.736 ns  max 1.202 ns  aver 1.040 ns
dim 513: min 0.681 ns  max 1.052 ns  aver 0.910 ns
dim 582: min 0.588 ns  max 0.949 ns  aver 0.785 ns
dim 659: min 0.468 ns  max 0.856 ns  aver 0.650 ns
dim 746: min 0.417 ns  max 0.809 ns  aver 0.601 ns
dim 844: min 0.356 ns  max 0.743 ns  aver 0.568 ns
dim 954: min 0.325 ns  max 0.696 ns  aver 0.529 ns
dim 1078: min 0.342 ns  max 0.670 ns  aver 0.510 ns
dim 1217: min 0.313 ns  max 0.614 ns  aver 0.460 ns
dim 1374: min 0.291 ns  max 0.581 ns  aver 0.433 ns
dim 1550: min 0.288 ns  max 0.562 ns  aver 0.421 ns
dim 1748: min 0.276 ns  max 0.525 ns  aver 0.405 ns
dim 1971: min 0.259 ns  max 0.513 ns  aver 0.383 ns
```

So what do the numbers mean? Is it supposed to be the average number of nanoseconds per operation?

**Comment by wbhart** [comment]
I am sceptical that tuning this way is a good idea. There are going to be cutoffs that depend on both the number of bits and the dimension. I see here timings only per dimension. You won't be able to come up with good tuning cutoff hyperplanes in the tesseract if you do this.

**Comment by wbhart** [comment]
By the way, the test code passes with NMOD_MAT_MUL_OPTIMAL_BITS set to 20, so presumably we just need to prove the code is correct. If not we can write our own 3x2 reduction. There is a paper on it somewhere.

File: fmpz/get_nmod.c (hunk 1 above)
Comment (tthsqe12) [reply fmpz/get_nmod.c:14]: Doesn't call %

**Comment by tthsqe12** [comment]
> I am sceptical that tuning this way is a good idea.

I am not tuning anything, just verifying that threads help.

> By the way, the test code passes with NMOD_MAT_MUL_OPTIMAL_BITS set to 20, so presumably we just need to prove the code is correct. If not we can write our own 3x2 reduction. There is a paper on it somewhere.

I have also fixed the M[1]!=0 assumption, and there is no need for a 3x2 reduction anymore.

File: fmpz/get_nmod.c (hunk 1 above)
Comment (wbhart) [reply fmpz/get_nmod.c:14]: So it's faster? If so why not speed up the other function so everything benefits, rather than create a new one?

Comment (tthsqe12) [reply fmpz/get_nmod.c:14]: What other functions? fmpz_fdiv_ui doesn't take a precomputed inverse...

Comment (wbhart) [reply fmpz/get_nmod.c:14]: Oh I see

**Comment by tthsqe12** [comment]
From my point of view, this is good to go now.

**Comment by fredrik-johansson** [comment]
Looks great.

**Pull request merged**